[#34] 책 카테고리 리스트로 조회하는 기능 추가 (feature/34-fix는 해당 이슈와 맞지 않는 기능이 삭제된 브랜치)

thinker-94 opened the pull request:
#34의 이슈와는 관련없는 리팩토링을 진행해서 fix 브랜치를 따로 만들어서 pr 날립니다.

Commit:
[#34] 책 카테고리 리스트로 조회하는 기능 추가

Commit:
bookQuery.xml 주석 리뷰 반영

카테고리의 개수가 많을 수 있다는 것 처럼 느껴질 수 있는 주석 수정

f-lab-dev commented:
- PR제목같은 디테일도 계속 신경써주세요
- 어떤 것을 어떻게 리팩토링했는지도 써주세요~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/BookController.java:
```diff
@@ -7,21 +7,29 @@
+   * @return 모든 책의 카테고리를 리턴한다 ex) 인문학, 철학 ..
+   */
+  @GetMapping
+  public List<BookDto> getCategoryList() {
```
메소드 이름도 `getAllCategoryList` 이런식으로 지어줘야 모든 것을 가져온다는 것을 유추할 수 있을겁니다~

thinker-94 replied:
네 구체적으로 작성하겠습니다.

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/implementation/BookServiceImpl.java:
```diff
@@ -27,6 +24,13 @@
 
+  //TODO. 캐시 key 생성
+//  @Cacheable(cacheNames = "bookCategoryList")
+  public List<BookDto> cacheBookCategoryList() {
```
이 메소드의 관심사는 카테고리의 리스트를 가져오는 것이지 캐시같은 구체적인 구현내용을 메소드 이름에 드러낼 필요가 있을까요?

thinker-94 replied:
네 구체적인 구현내용은 메소드명에 들어갈 필요가 없다 생각합니다. 그리고 비즈니스 요구사항에 캐싱은 없기 때문에 해당 메소드명은 문제가 있습니다

thinker-94 commented:
> * PR제목같은 디테일도 계속 신경써주세요
> * 어떤 것을 어떻게 리팩토링했는지도 써주세요~

해당 리팩토링은 따로 이슈 만들어서 진행하려고 합니다. 그래도 작성 하는게 좋을까요?

Commit:
[#34] BookController, BookService 메소드명 변경, 주석 추가

Service Layer는 비즈니스 요구사항이 메소드 이름으로 표현되는게 좋겠다고 생각했다. 그래서 show 라는 요구사항이 표현되는 이름으로 변경

현 프로젝트의 Controller Layer는 REST 규약에 맞게 클라이언트에게 데이터를 받고 전달하는 역할을 한다. 그래서 HTTP METHOD + 가져오는 데이터의 양을 표현하는 명사 + 해당 도메인(book)에 관련된 이름으로 메소드 명 규칙을 정함

Mapper 인터페이스의 메소드명은 개발 용어 위주로 메소드명을 정하기로함. 그래서 All보다는 List처럼 자료구조 이름을 넣음

thinker-94 commented:
레이어별로 메소드명 규칙을 정해서 문서화를 하고 싶은데, 문서는 위키에 추가하는게 어떨까요? 

메소드 명이 일관성이 없어서 기능 추가와 유지보수에 문제가 생길 것 같다는 생각이 들었기 때문입니다. 이유는 메소드명을 어느정도 정해진 규칙없이 생각하고 이해하는 시간이 늘어난다고 판단했기 때문입니다. 그래서 레이어별로 메소드명 규칙을 정하고 문서화를 하려고 합니다

실제로 개발 중에 메소드명을 고민하는데 시간 소요가 많이 되고 있습니다.

f-lab-dev commented:
컨벤션을 만든다 하더라도 설계하다보면 각 메소드별로 다 역할이 다를텐데 그것을 컨벤션만으로 해결할 수 있을까요?
어느정도의 컨벤션은 있으면 좋지만 모든 케이스를 일반화하려다보면 어려울수도 있습니다.
메소드의 역할과 시그니쳐를 정하는것은 어느 정도의 고민이 필요한 부분이긴 합니다

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/BookController.java:
```diff
@@ -26,8 +26,8 @@
-  public List<BookDto> getCategoryList() {
-    return bookService.cacheBookCategoryList();
+  public List<BookDto> getAllCategory() {
+    return bookService.showAllBookCategory();
```
`show`는 보여준다인데 백엔드에서 뭔가를 보여줄 일이 있을까요?

thinker-94 replied:
백엔드 기준으로 생각하진 않았어요. 비즈니스 로직은 비즈니스 요구사항이라는 표현이 메소드명에 들어가는게 좋다고 생각했기 때문입니다. 기획자에게 가끔 비즈니스 로직에 대해서 설명하는 경우가 있다는걸 들어서 이름을 show로 해봤습니다. 

저 메소드는 비즈니스 요구사항이 책 카테고리 목록을 사용자에게 보여준다는것을 나타내는 메소드입니다. 그런데 show라는게 프론트엔드에서 view를 보여주는 느낌이 날 수도 있을것 같네요. 나은 이름으로 생각해보겠습니다

thinker-94 commented:
> 컨벤션을 만든다 하더라도 설계하다보면 각 메소드별로 다 역할이 다를텐데 그것을 컨벤션만으로 해결할 수 있을까요?
> 어느정도의 컨벤션은 있으면 좋지만 모든 케이스를 일반화하려다보면 어려울수도 있습니다.
> 메소드의 역할과 시그니쳐를 정하는것은 어느 정도의 고민이 필요한 부분이긴 합니다

진도가 중요하다 생각해서 controller service repository의 대략적인 규칙만 정했습니다.

Commit:
[#38] 가독성을 위해 show에서 give로 메소드 동사 변경

f-lab-dev commented:
컨벤션과 진도의 연관성에 대해 이해가 되질 않는데요. 보충설명 부탁드립니다~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/implementation/BookServiceImpl.java:
```diff
@@ -26,7 +26,7 @@
   //TODO. 캐시 key 생성
 //  @Cacheable(cacheNames = "bookCategoryList")
-  public List<BookDto> showAllBookCategory() {
+  public List<BookDto> giveAllBookCategory() {
```
give도 일반적으로 이런 상황에 사용되지 않는데 get이 나을 것 같네요
